filebeat/input/state.go

```diff
@@ -20,3 +26,76 @@ func NewFileState(fileInfo os.FileInfo, path string) FileState {
 		FileStateOS: GetOSFileState(fileInfo),
 	}
 }
+
+// States handles list of FileState
+type States struct {
+	states []FileState
+	mutex  sync.Mutex
+}
+
+func NewStates() *States {
+	return &States{
+		states: []FileState{},
+	}
+}
+
+// Update updates a state. If previous state didn't exist, new one is created
+func (s *States) Update(newState FileState) {
+
+	index, oldState := s.FindPrevious(newState)
+
+	// Only block after FindPrevious as FindPrevious has its own lock
```

Comment:
is there a chance for a race between FindPrevious and actual body updating the state?